Test suite failure with GHC 8

```
> /tmp/stackage-build9/xmonad-0.12$ runghc -clear-package-db -global-package-db -package-db=/home/stackage/work/builds/nightly/pkgdb Setup configure --enable-tests --package-db=clear --package-db=global --package-db=/home/stackage/work/builds/nightly/pkgdb --libdir=/home/stackage/work/builds/nightly/lib --bindir=/home/stackage/work/builds/nightly/bin --datadir=/home/stackage/work/builds/nightly/share --libexecdir=/home/stackage/work/builds/nightly/libexec --sysconfdir=/home/stackage/work/builds/nightly/etc --docdir=/home/stackage/work/builds/nightly/doc/xmonad-0.12 --htmldir=/home/stackage/work/builds/nightly/doc/xmonad-0.12 --haddockdir=/home/stackage/work/builds/nightly/doc/xmonad-0.12 --flags=
Configuring xmonad-0.12...
Warning: 'ghc-options: -prof' is not necessary and will lead to problems when
used on a library. Use the configure flag --enable-library-profiling and/or
--enable-profiling.
> /tmp/stackage-build9/xmonad-0.12$ runghc -clear-package-db -global-package-db -package-db=/home/stackage/work/builds/nightly/pkgdb Setup build
Building xmonad-0.12...
Preprocessing library xmonad-0.12...
[1 of 9] Compiling Paths_xmonad     ( dist/build/autogen/Paths_xmonad.hs, dist/build/Paths_xmonad.o )
[2 of 9] Compiling XMonad.StackSet  ( src/XMonad/StackSet.hs, dist/build/XMonad/StackSet.o )

src/XMonad/StackSet.hs:485:1: warning: [-Wredundant-constraints]
    • Redundant constraint: Eq s
    • In the type signature for:
           delete' :: (Eq a, Eq s) =>
                      a -> StackSet i l a s sd -> StackSet i l a s sd

src/XMonad/StackSet.hs:550:1: warning: [-Wredundant-constraints]
    • Redundant constraint: Eq a
    • In the type signature for:
           shiftWin :: (Ord a, Eq a, Eq s, Eq i) =>
                       i -> a -> StackSet i l a s sd -> StackSet i l a s sd
[3 of 9] Compiling XMonad.Core      ( src/XMonad/Core.hs, dist/build/XMonad/Core.o )

src/XMonad/Core.hs:37:1: warning: [-Wunused-imports]
    The import of ‘Control.Applicative’ is redundant
      except perhaps to import instances from ‘Control.Applicative’
    To import instances alone, use: import Control.Applicative()
[4 of 9] Compiling XMonad.Layout    ( src/XMonad/Layout.hs, dist/build/XMonad/Layout.o )

src/XMonad/Layout.hs:140:1: warning: [-Wredundant-constraints]
    • Redundant constraints: (LayoutClass l a, LayoutClass r a)
    • In the type signature for:
           (|||) :: (LayoutClass l a, LayoutClass r a) =>
                    l a -> r a -> Choose l r a
[5 of 9] Compiling XMonad.Operations ( src/XMonad/Operations.hs, dist/build/XMonad/Operations.o )

src/XMonad/Operations.hs:32:1: warning: [-Wunused-imports]
    The import of ‘Control.Applicative’ is redundant
      except perhaps to import instances from ‘Control.Applicative’
    To import instances alone, use: import Control.Applicative()
[6 of 9] Compiling XMonad.ManageHook ( src/XMonad/ManageHook.hs, dist/build/XMonad/ManageHook.o )
[7 of 9] Compiling XMonad.Config    ( src/XMonad/Config.hs, dist/build/XMonad/Config.o )
[8 of 9] Compiling XMonad.Main      ( src/XMonad/Main.hs, dist/build/XMonad/Main.o )
[9 of 9] Compiling XMonad           ( src/XMonad.hs, dist/build/XMonad.o )
Preprocessing executable 'xmonad' for xmonad-0.12...
[1 of 1] Compiling Main             ( Main.hs, dist/build/xmonad/xmonad-tmp/Main.o )
Linking dist/build/xmonad/xmonad ...
Preprocessing test suite 'properties' for xmonad-0.12...
[ 1 of 18] Compiling Properties.Failure ( tests/Properties/Failure.hs, dist/build/properties/properties-tmp/Properties/Failure.o )
[ 2 of 18] Compiling Utils            ( tests/Utils.hs, dist/build/properties/properties-tmp/Utils.o )
[ 3 of 18] Compiling Instances        ( tests/Instances.hs, dist/build/properties/properties-tmp/Instances.o )
[ 4 of 18] Compiling Properties.Floating ( tests/Properties/Floating.hs, dist/build/properties/properties-tmp/Properties/Floating.o )
[ 5 of 18] Compiling Properties.Stack ( tests/Properties/Stack.hs, dist/build/properties/properties-tmp/Properties/Stack.o )
[ 6 of 18] Compiling Properties.Layout.Full ( tests/Properties/Layout/Full.hs, dist/build/properties/properties-tmp/Properties/Layout/Full.o )
[ 7 of 18] Compiling Properties.Delete ( tests/Properties/Delete.hs, dist/build/properties/properties-tmp/Properties/Delete.o )
[ 8 of 18] Compiling Properties.Focus ( tests/Properties/Focus.hs, dist/build/properties/properties-tmp/Properties/Focus.o )
[ 9 of 18] Compiling Properties.GreedyView ( tests/Properties/GreedyView.hs, dist/build/properties/properties-tmp/Properties/GreedyView.o )
[10 of 18] Compiling Properties.Insert ( tests/Properties/Insert.hs, dist/build/properties/properties-tmp/Properties/Insert.o )
[11 of 18] Compiling Properties.Screen ( tests/Properties/Screen.hs, dist/build/properties/properties-tmp/Properties/Screen.o )
[12 of 18] Compiling Properties.Shift ( tests/Properties/Shift.hs, dist/build/properties/properties-tmp/Properties/Shift.o )
[13 of 18] Compiling Properties.StackSet ( tests/Properties/StackSet.hs, dist/build/properties/properties-tmp/Properties/StackSet.o )
[14 of 18] Compiling Properties.Swap  ( tests/Properties/Swap.hs, dist/build/properties/properties-tmp/Properties/Swap.o )
[15 of 18] Compiling Properties.View  ( tests/Properties/View.hs, dist/build/properties/properties-tmp/Properties/View.o )
[16 of 18] Compiling Properties.Workspace ( tests/Properties/Workspace.hs, dist/build/properties/properties-tmp/Properties/Workspace.o )
[17 of 18] Compiling Properties.Layout.Tall ( tests/Properties/Layout/Tall.hs, dist/build/properties/properties-tmp/Properties/Layout/Tall.o )
[18 of 18] Compiling Main             ( tests/Properties.hs, dist/build/properties/properties-tmp/Main.o )
Linking dist/build/properties/properties ...
> /tmp/stackage-build9/xmonad-0.12$ dist/build/properties/properties
StackSet invariants                : +++ OK, passed 100 tests.
empty: invariant                   : +++ OK, passed 100 tests.
empty is empty                     : +++ OK, passed 100 tests.
empty / current                    : +++ OK, passed 100 tests.
empty / member                     : +++ OK, passed 100 tests.
view : invariant                   : +++ OK, passed 100 tests.
view sets current                  : +++ OK, passed 100 tests.
view idempotent                    : +++ OK, passed 100 tests.
view reversible                    : +++ OK, passed 100 tests.
view is local                      : +++ OK, passed 100 tests.
greedyView : invariant             : +++ OK, passed 100 tests.
greedyView sets current            : +++ OK, passed 100 tests.
greedyView is safe                 : +++ OK, passed 100 tests.
greedyView idempotent              : +++ OK, passed 100 tests.
greedyView reversible              : +++ OK, passed 100 tests.
greedyView is local                : +++ OK, passed 100 tests.
peek/member                        : +++ OK, passed 100 tests.
index/length                       : +++ OK, passed 100 tests.
focus left : invariant             : +++ OK, passed 100 tests.
focus master : invariant           : +++ OK, passed 100 tests.
focus right: invariant             : +++ OK, passed 100 tests.
focusWindow: invariant             : +++ OK, passed 100 tests.
focus left/master                  : +++ OK, passed 100 tests.
focus right/master                 : +++ OK, passed 100 tests.
focus master/master                : +++ OK, passed 100 tests.
focusWindow master                 : +++ OK, passed 100 tests.
focus left/right                   : +++ OK, passed 100 tests.
focus right/left                   : +++ OK, passed 100 tests.
focus all left                     : +++ OK, passed 100 tests.
focus all right                    : +++ OK, passed 100 tests.
focus down is local                : +++ OK, passed 100 tests.
focus up is local                  : +++ OK, passed 100 tests.
focus master is local              : +++ OK, passed 100 tests.
focus master idemp                 : +++ OK, passed 100 tests.
focusWindow is local               : +++ OK, passed 100 tests.
focusWindow works                  : +++ OK, passed 100 tests.
focusWindow identity               : +++ OK, passed 100 tests.
findTag                            : +++ OK, passed 100 tests.
allWindows/member                  : +++ OK, passed 100 tests.
currentTag                         : +++ OK, passed 100 tests.
insert: invariant                  : +++ OK, passed 100 tests.
insert/new                         : +++ OK, passed 100 tests.
insert is idempotent               : +++ OK, passed 100 tests.
insert is reversible               : +++ OK, passed 100 tests.
insert is local                    : +++ OK, passed 100 tests.
insert duplicates                  : +++ OK, passed 100 tests.
insert/peek                        : +++ OK, passed 100 tests.
insert/size                        : +++ OK, passed 100 tests.
delete: invariant                  : +++ OK, passed 100 tests.
delete/empty                       : +++ OK, passed 100 tests.
delete/member                      : +++ OK, passed 100 tests.
delete is reversible               : +++ OK, passed 100 tests.
delete is local                    : +++ OK, passed 100 tests.
delete/focus                       : +++ OK, passed 100 tests.
delete  last/focus up              : +++ OK, passed 100 tests.
delete ~last/focus down            : +++ OK, passed 100 tests.
filter preserves order             : +++ OK, passed 100 tests.
swapLeft                           : +++ OK, passed 100 tests.
swapRight                          : +++ OK, passed 100 tests.
swapMaster: invariant              : +++ OK, passed 100 tests.
swapUp: invariant                  : +++ OK, passed 100 tests.
swapDown: invariant                : +++ OK, passed 100 tests.
swapMaster id on focus             : +++ OK, passed 100 tests.
swapUp id on focus                 : +++ OK, passed 100 tests.
swapDown id on focus               : +++ OK, passed 100 tests.
swapMaster is idempotent           : +++ OK, passed 100 tests.
swap all left                      : +++ OK, passed 100 tests.
swap all right                     : +++ OK, passed 100 tests.
swapMaster is local                : +++ OK, passed 100 tests.
swapUp is local                    : +++ OK, passed 100 tests.
swapDown is local                  : +++ OK, passed 100 tests.
shiftMaster id on focus            : +++ OK, passed 100 tests.
shiftMaster is local               : +++ OK, passed 100 tests.
shiftMaster is idempotent          : +++ OK, passed 100 tests.
shiftMaster preserves ordering     : +++ OK, passed 100 tests.
shift: invariant                   : +++ OK, passed 100 tests.
shift is reversible                : +++ OK, passed 100 tests.
shiftWin: invariant                : +++ OK, passed 100 tests.
shiftWin is shift on focus         : +++ OK, passed 100 tests.
shiftWin fix current               : +++ OK, passed 100 tests.
shiftWin identity                  : +++ OK, passed 100 tests.
floating is reversible             : +++ OK, passed 100 tests.
floating sets geometry             : +++ OK, passed 100 tests.
floats can be deleted              : +++ OK, passed 100 tests.
screens includes current           : +++ OK, passed 100 tests.
differentiate works                : +++ OK, passed 100 tests.
lookupTagOnScreen                  : +++ OK, passed 100 tests.
lookupTagOnVisbleScreen            : +++ OK, passed 100 tests.
screens works                      : +++ OK, passed 100 tests.
renaming works                     : +++ OK, passed 100 tests.
ensure works                       : +++ OK, passed 100 tests.
ensure hidden semantics            : +++ OK, passed 100 tests.
mapWorkspace id                    : +++ OK, passed 100 tests.
mapWorkspace inverse               : +++ OK, passed 100 tests.
mapLayout id                       : +++ OK, passed 100 tests.
mapLayout inverse                  : +++ OK, passed 100 tests.
abort fails                        : *** Failed! Falsifiable (after 1 test): 
0
new fails with abort               : *** Failed! Falsifiable (after 1 test): 
0
point within                       : +++ OK, passed 100 tests.
tile 1 window fullsize             : +++ OK, passed 100 tests.
tiles never overlap                : +++ OK, passed 100 tests.
split horizontal                   : +++ OK, passed 100 tests.
split vertical                     : +++ OK, passed 100 tests.
pure layout tall                   : +++ OK, passed 100 tests.
send shrink    tall                : +++ OK, passed 100 tests.
send expand    tall                : +++ OK, passed 100 tests.
send incmaster tall                : +++ OK, passed 100 tests.
pure layout full                   : +++ OK, passed 100 tests.
send message full                  : +++ OK, passed 100 tests.
describe full                      : +++ OK, passed 100 tests.
describe mirror                    : +++ OK, passed 100 tests.
window resize hints: inc           : +++ OK, passed 100 tests.
window resize hints: inc all       : +++ OK, passed 100 tests.
window resize hints: max           : +++ OK, passed 100 tests.
window resize hints: max all       : +++ OK, passed 100 tests.
window aspect hints: fits          : +++ OK, passed 100 tests.
window aspect hints: shrinks       : +++ OK, passed 100 tests.
pointWithin                        : +++ OK, passed 100 tests.
pointWithin mirror                 : +++ OK, passed 100 tests.
properties: 2 test(s) failed
CallStack (from HasCallStack):
  error, called at tests/Properties.hs:39:20 in main:Main
```

--- a/src/XMonad/Layout.hs
+++ b/src/XMonad/Layout.hs
@@ -137,7 +137,7 @@ data ChangeLayout = FirstLayout | NextLayout deriving (Eq, Show, Typeable)
 instance Message ChangeLayout
 
 -- | The layout choice combinator
-(|||) :: (LayoutClass l a, LayoutClass r a) => l a -> r a -> Choose l r a
+(|||) :: l a -> r a -> Choose l r a
 (|||) = Choose L
 infixr 5 |||
 

--- a/src/XMonad/StackSet.hs
+++ b/src/XMonad/StackSet.hs
@@ -477,12 +477,12 @@ insertUp a s = if member a s then s else insert
 --
 --   * otherwise, delete doesn't affect the master.
 --
-delete :: (Ord a, Eq s) => a -> StackSet i l a s sd -> StackSet i l a s sd
+delete :: (Ord a) => a -> StackSet i l a s sd -> StackSet i l a s sd
 delete w = sink w . delete' w
 
 -- | Only temporarily remove the window from the stack, thereby not destroying special
 -- information saved in the 'Stackset'
-delete' :: (Eq a, Eq s) => a -> StackSet i l a s sd -> StackSet i l a s sd
+delete' :: (Eq a) => a -> StackSet i l a s sd -> StackSet i l a s sd
 delete' w s = s { current = removeFromScreen        (current s)
                 , visible = map removeFromScreen    (visible s)
                 , hidden  = map removeFromWorkspace (hidden  s) }
@@ -547,7 +547,7 @@ shift n s = maybe s (\w -> shiftWin n w s) (peek s)
 -- focused element on that workspace.
 -- The actual focused workspace doesn't change. If the window is not
 -- found in the stackSet, the original stackSet is returned.
-shiftWin :: (Ord a, Eq a, Eq s, Eq i) => i -> a -> StackSet i l a s sd -> StackSet i l a s sd
+shiftWin :: (Ord a, Eq s, Eq i) => i -> a -> StackSet i l a s sd -> StackSet i l a s sd
 shiftWin n w s = case findTag w s of
                     Just from | n `tagMember` s && n /= from -> go from s
                     _                                        -> s

--- a/tests/Properties/Failure.hs
+++ b/tests/Properties/Failure.hs
@@ -4,23 +4,27 @@ import XMonad.StackSet hiding (filter)
 
 import qualified Control.Exception.Extensible as C
 import System.IO.Unsafe
+import Data.List (isPrefixOf)
 
 -- ---------------------------------------------------------------------
--- testing for failure
-
--- and help out hpc
-prop_abort x = unsafePerformIO $ C.catch (abort "fail")
-                                         (\(C.SomeException e) -> return $  show e == "xmonad: StackSet: fail" )
+-- testing for failure and help out hpc
+--
+-- Since base 4.9.0.0 `error` appends a stack trace. The tests below
+-- use `isPrefixOf` to only test equality on the error message.
+--
+prop_abort :: Int -> Bool
+prop_abort _ = unsafePerformIO $ C.catch (abort "fail") check
    where
-     _ = x :: Int
+     check (C.SomeException e) =
+       return $ "xmonad: StackSet: fail" `isPrefixOf` show e
 
 -- new should fail with an abort
-prop_new_abort x = unsafePerformIO $ C.catch f
-                                         (\(C.SomeException e) -> return $ show e == "xmonad: StackSet: non-positive argument to StackSet.new" )
+prop_new_abort :: Int -> Bool
+prop_new_abort _ = unsafePerformIO $ C.catch f check
    where
      f = new undefined{-layout-} [] [] `seq` return False
-
-     _ = x :: Int
+     check (C.SomeException e) =
+       return $ "xmonad: StackSet: non-positive argument to StackSet.new" `isPrefixOf` show e
 
 -- TODO: Fix this?
 -- prop_view_should_fail = view {- with some bogus data -}

--- a/xmonad.cabal
+++ b/xmonad.cabal
@@ -31,7 +31,8 @@ build-type:         Simple
 tested-with:
                     GHC==7.6.3,
                     GHC==7.8.4,
-                    GHC==7.10.2
+                    GHC==7.10.3,
+                    GHC==8.0.1
 
 data-files:         man/xmonad.hs, man/xmonad.1, man/xmonad.1.html
 

